File: core/src/main/java/dev/vibris/core/EngineShutdown.java
package dev.vibris.core;

import dev.vibris.api.VibrisRuntimeAdapter;

import java.util.concurrent.ScheduledExecutorService;

final class EngineShutdown {
    private EngineShutdown() {
    }

    static void close(
        VibrisRuntimeAdapter runtime,
        ScheduledExecutorService disconnectTimer,
        FairJobScheduler scheduler,
        SourceActivator activator
    ) {
        RuntimeException failure = null;
        try {
            runtime.close();
        } catch (RuntimeException exception) {
            failure = exception;
        }
        disconnectTimer.shutdownNow();
        try {
            scheduler.close();
        } catch (RuntimeException exception) {
            if (failure == null) failure = exception;
            else failure.addSuppressed(exception);
        } finally {
            activator.close();
        }
        if (failure != null) throw failure;
    }
}

File: core/src/main/kotlin/dev/vibris/core/EngineShutdown.kt
package dev.vibris.core

import dev.vibris.api.VibrisRuntimeAdapter
import java.util.concurrent.ScheduledExecutorService

internal object EngineShutdown {
    @JvmStatic
    fun close(
        runtime: VibrisRuntimeAdapter,
        disconnectTimer: ScheduledExecutorService,
        scheduler: FairJobScheduler,
        activator: SourceActivator,
    ) {
        var failure: RuntimeException? = null
        try {
            runtime.close()
        } catch (exception: RuntimeException) {
            failure = exception
        }
        disconnectTimer.shutdownNow()
        try {
            scheduler.close()
        } catch (exception: RuntimeException) {
            if (failure == null) {
                failure = exception
            } else {
                failure.addSuppressed(exception)
            }
        } finally {
            activator.close()
        }
        if (failure != null) {
            throw failure
        }
    }
}